File: static/js/turnover.js
async function tvLoad(view) {
  spin('tvContent');
  try {
    if (view === 'compare') {
      const d = await api('/api/turnover/compare-models');
      const models = d.models||{}, best = d.best_model||'';
      const mNames = {random_forest:'Random Forest',logistic_regression:'Logistic Regression',xgboost:'XGBoost (GBM)'};
      const mCols  = {random_forest:'var(--green)',logistic_regression:'var(--blue)',xgboost:'var(--amber)'};
      set('tvContent', `
        <div class="alert alert-green"><span>⚡</span><div class="body">${d.recommendation}</div></div>
        <div class="grid-3">${Object.entries(models).map(([name,m])=>`
          <div class="model-card ${name===best?'best':''}">
            <div class="model-name" style="color:${mCols[name]}">${name===best?'⭐ ':''} ${mNames[name]||name}</div>
            <div class="model-metrics">
              <div class="mm"><div class="mm-val" style="color:${mCols[name]}">${(m.accuracy*100).toFixed(1)}%</div><div class="mm-label">Accuracy</div></div>
              <div class="mm"><div class="mm-val" style="color:${mCols[name]}">${m.roc_auc}</div><div class="mm-label">AUC</div></div>
              <div class="mm"><div class="mm-val" style="color:${mCols[name]}">${(m.f1_score*100).toFixed(1)}%</div><div class="mm-label">F1</div></div>
              <div class="mm"><div class="mm-val" style="color:${mCols[name]}">${m.cv5_auc}</div><div class="mm-label">CV-5</div></div>
            </div>
            <div style="margin-top:10px;font-size:11px;color:var(--muted)">Top feature: <strong style="color:var(--text);font-family:var(--mono)">${m.top_feature}</strong></div>
          </div>`).join('')}
        </div>
        <div class="grid-2">
          <div class="card"><div class="card-title">ROC AUC Comparison</div><div class="chart-box" style="height:180px"><canvas id="tvAucChart"></canvas></div></div>
          <div class="card"><div class="card-title">Accuracy Comparison</div><div class="chart-box" style="height:180px"><canvas id="tvAccChart"></canvas></div></div>
        </div>`);
      setTimeout(()=>{
        const names=Object.keys(models).map(k=>mNames[k]), cols=Object.keys(models).map(k=>mCols[k].replace('var(--green)','rgba(0,229,160,.7)').replace('var(--blue)','rgba(56,189,248,.7)').replace('var(--amber)','rgba(245,158,11,.7)'));
        mkChart('tvAucChart','bar',names,[{data:Object.values(models).map(m=>m.roc_auc),backgroundColor:cols,borderRadius:4}],{scales:{y:{min:0.7,max:1,ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},x:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
        mkChart('tvAccChart','bar',names,[{data:Object.values(models).map(m=>(m.accuracy*100).toFixed(1)),backgroundColor:cols,borderRadius:4}],{scales:{y:{min:70,ticks:{callback:v=>v+'%',color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},x:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
      },50);

    } else if (view === 'train') {
      const model = $('tvModel').value;
      const d = await api(`/api/turnover/train?model_type=${model}&top_n_employees=20`);
      const m = d.metrics, fi = Object.entries(d.feature_importance||{}).slice(0,8);
      const cm = d.confusion_matrix||{}, seg = d.risk_segments||{};
      const emp = d.top_risk_employees||[], ins = d.auto_insights||[];
      const dist = d.attrition_prob_distribution||{};
      const rCol = p => p>=0.75?'var(--red)':p>=0.5?'var(--amber)':p>=0.25?'var(--blue)':'var(--green)';
      set('tvContent', `
        ${ins.map(i=>`<div class="alert alert-green" style="margin-bottom:6px"><span>↺</span><div class="body">${i}</div></div>`).join('')}
        <div class="grid-4">
          <div class="stat"><div class="stat-label">Accuracy</div><div class="stat-value green">${(m.accuracy*100).toFixed(1)}%</div></div>
          <div class="stat"><div class="stat-label">ROC AUC</div><div class="stat-value blue">${m.roc_auc}</div></div>
          <div class="stat"><div class="stat-label">F1 Score</div><div class="stat-value purple">${(m.f1_score*100).toFixed(1)}%</div></div>
          <div class="stat"><div class="stat-label">CV-5 AUC</div><div class="stat-value amber">${m.cv5_auc}</div></div>
        </div>
        <div class="grid-4">
          <div class="stat"><div class="stat-label">Critical Risk ≥75%</div><div class="stat-value red">${seg.critical||0}</div></div>
          <div class="stat"><div class="stat-label">High Risk 50–75%</div><div class="stat-value amber">${seg.high||0}</div></div>
          <div class="stat"><div class="stat-label">Medium Risk 25–50%</div><div class="stat-value blue">${seg.medium||0}</div></div>
          <div class="stat"><div class="stat-label">Low Risk &lt;25%</div><div class="stat-value green">${seg.low||0}</div></div>
        </div>
        <div class="grid-2">
          <div class="card"><div class="card-title">Feature Importance</div>${fi.map(([n,v])=>`<div class="bar-row"><div class="bar-label">${n}</div><div class="bar-track"><div class="bar-fill blue" style="width:${v/fi[0][1]*100}%"></div></div><div class="bar-val">${v}</div></div>`).join('')}</div>
          <div class="card"><div class="card-title">Confusion Matrix</div>
            <div class="cm-grid">
              <div class="cm-cell" style="background:rgba(0,229,160,.08);border-color:rgba(0,229,160,.2)"><div class="cm-val" style="color:var(--green)">${cm.true_negatives}</div><div class="cm-label">True Negatives<br>Correctly stayed</div></div>
              <div class="cm-cell" style="background:rgba(248,113,113,.07);border-color:rgba(248,113,113,.2)"><div class="cm-val" style="color:var(--red)">${cm.false_positives}</div><div class="cm-label">False Positives<br>Predicted left, stayed</div></div>
              <div class="cm-cell" style="background:rgba(245,158,11,.07);border-color:rgba(245,158,11,.2)"><div class="cm-val" style="color:var(--amber)">${cm.false_negatives}</div><div class="cm-label">False Negatives<br>Missed leavers</div></div>
              <div class="cm-cell" style="background:rgba(56,189,248,.07);border-color:rgba(56,189,248,.2)"><div class="cm-val" style="color:var(--blue)">${cm.true_positives}</div><div class="cm-label">True Positives<br>Correctly predicted left</div></div>
            </div>
          </div>
        </div>
        <div class="card"><div class="card-title">Attrition Probability Distribution</div><div class="chart-box" style="height:180px"><canvas id="tvDistChart"></canvas></div></div>
        <div class="card">
          <div class="card-title">🔥 Highest Risk Employees</div>
          <div class="tbl-wrap"><table>
            <thead><tr><th>ID</th><th>Role</th><th>Salary</th><th>Satisfaction</th><th>Hrs/month</th><th>Attrition Prob</th><th>Risk</th><th>Actual</th></tr></thead>
            <tbody>${emp.slice(0,15).map(e=>`<tr>
              <td class="val">${e.index}</td>
              <td style="font-size:11px">${e.job_role}</td>
              <td>${badge(e.salary, e.salary==='low'?'red':e.salary==='medium'?'amber':'green')}</td>
              <td class="${e.satisfaction_level<0.4?'danger':'val'}">${e.satisfaction_level}</td>
              <td class="${e.average_montly_hours>250?'warn':'val'}">${e.average_montly_hours}</td>
              <td style="font-weight:600;color:${rCol(e.attrition_probability)};font-family:var(--mono)">${(e.attrition_probability*100).toFixed(1)}%</td>
              <td>${badge(e.risk_label, e.risk_label==='Critical'?'red':e.risk_label==='High'?'amber':'blue')}</td>
              <td style="color:${e.actual_left?'var(--red)':'var(--green)'}">${e.actual_left?'Left':'Stayed'}</td>
            </tr>`).join('')}</tbody>
          </table></div>
        </div>
        <div class="card"><div class="card-title">Retention Recommendations</div>${(d.retention_recommendations||[]).map(r=>`<div style="padding:9px 0;border-bottom:1px solid var(--border);font-size:12px;color:var(--muted)">→ ${r}</div>`).join('')}</div>`);
      setTimeout(()=>{
        const L=Object.keys(dist),V=Object.values(dist);
        mkChart('tvDistChart','bar',L,[{data:V,backgroundColor:L.map((_,i)=>i<4?'rgba(0,229,160,.6)':i<7?'rgba(245,158,11,.6)':'rgba(248,113,113,.7)'),borderRadius:3}],{scales:{x:{ticks:{color:'#6b7280',maxRotation:30,font:{size:9,family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},y:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
      },50);

    } else if (view === 'eda') {
      const d = await api('/api/turnover/eda');
      const byRole = d.turnover_rate_by_role_pct||{}, bySal = d.turnover_rate_by_salary_pct||{}, satRole = d.satisfaction_by_role||{};
      set('tvContent', `
        <div class="grid-3">
          <div class="stat"><div class="stat-label">Overall Turnover</div><div class="stat-value red">${d.overall_turnover_rate_pct}%</div></div>
          <div class="stat"><div class="stat-label">Satisfaction (left)</div><div class="stat-value amber">${d.avg_satisfaction_who_left}</div><div class="stat-sub">vs ${d.avg_satisfaction_who_stayed} who stayed</div></div>
          <div class="stat"><div class="stat-label">Avg Tenure (left)</div><div class="stat-value blue">${d.avg_tenure_years_who_left} yrs</div></div>
        </div>
        <div class="alert alert-amber"><span>↺</span><div class="body">${d.insight}</div></div>
        <div class="card"><div class="card-title">Turnover Rate by Role</div>
          ${Object.entries(byRole).sort((a,b)=>b[1]-a[1]).map(([role,pct])=>`<div class="risk-row">
            <div class="risk-name">${role.replace(/_/g,' ')}</div>
            <div class="risk-track"><div class="risk-fill" style="width:${pct}%;background:${rc(pct/100)}"></div></div>
            <div class="risk-pct" style="color:${rc(pct/100)}">${pct}%</div>
          </div>`).join('')}
        </div>
        <div class="grid-2">
          <div class="card"><div class="card-title">Turnover by Salary</div><div class="chart-box" style="height:200px"><canvas id="tvEdaSal"></canvas></div></div>
          <div class="card"><div class="card-title">Satisfaction by Role</div><div class="chart-box" style="height:200px"><canvas id="tvEdaSat"></canvas></div></div>
        </div>`);
      setTimeout(()=>{
        const sv=Object.values(bySal);
        mkChart('tvEdaSal','bar',Object.keys(bySal),[{data:sv,backgroundColor:sv.map(v=>rc(v/100)+'bb'),borderRadius:4}],{scales:{y:{ticks:{callback:v=>v+'%',color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},x:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
        mkChart('tvEdaSat','bar',Object.keys(satRole).map(k=>k.replace(/_/g,' ')),[{data:Object.values(satRole),backgroundColor:'rgba(167,139,250,.65)',borderRadius:4}],{indexAxis:'y',scales:{x:{ticks:{color:'#6b7280',font:{family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}},y:{ticks:{color:'#6b7280',font:{size:10,family:"'IBM Plex Mono'"}},grid:{color:'rgba(255,255,255,0.04)'}}}});
      },50);

    } else if (view === 'risk') {
      const d = await api('/api/turnover/risk-profile');
      const segs = d.turnover_risk_by_segment||{};
      set('tvContent', `
        <div class="stat" style="width:fit-content"><div class="stat-label">Overall Turnover</div><div class="stat-value red">${d.overall_turnover_pct}%</div></div>
        ${Object.entries(segs).map(([grp,vals])=>`<div class="card">
          <div class="card-title">Turnover by ${grp.replace(/_/g,' ')}</div>
          ${Object.entries(vals).sort((a,b)=>b[1]-a[1]).map(([k,v])=>`<div class="risk-row">
            <div class="risk-name">${k}</div>
            <div class="risk-track"><div class="risk-fill" style="width:${Math.min(v,100)}%;background:${rc(v/100)}"></div></div>
            <div class="risk-pct" style="color:${rc(v/100)}">${v}%</div>
          </div>`).join('')}
        </div>`).join('')}`);
    }
  } catch(e) { oops('tvContent', e); }
}

File: routers/turnover.py
"""
Employee Turnover Prediction
==============================
Three models: Logistic Regression, Random Forest, XGBoost (GradientBoosting)
Per-employee attrition probability, feature importance, SHAP-style insights.

Dataset: fau_clinic_turnover_data.csv
"""
import pandas as pd
import numpy as np
from fastapi import APIRouter, Query
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import (accuracy_score, precision_score, recall_score,
                             f1_score, roc_auc_score, confusion_matrix)
from sklearn.preprocessing import LabelEncoder
import warnings
warnings.filterwarnings("ignore")

from .utils import load_csv, df_info, corr_matrix

router    = APIRouter()
DEFAULT_FILE = "fau_clinic_turnover_data.csv"
TARGET    = "left"
SALARY_MAP = {"low": 0, "medium": 1, "high": 2}


def _preprocess(df: pd.DataFrame):
    d   = df.copy()
    log = {}
    le  = LabelEncoder()
    for col in d.select_dtypes(include="object").columns:
        d[col] = le.fit_transform(d[col].astype(str))
        log[col] = "LabelEncoded"
    d.fillna(d.median(numeric_only=True), inplace=True)
    return d, log


def _build_model(model_type: str):
    if model_type == "logistic_regression":
        return LogisticRegression(max_iter=500, random_state=42, C=1.0)
    elif model_type == "xgboost":
        return GradientBoostingClassifier(n_estimators=150, learning_rate=0.1, max_depth=4, random_state=42)
    else:
        return RandomForestClassifier(n_estimators=150, max_depth=None, random_state=42, n_jobs=-1)


def _feature_importance(model, feature_names: list) -> dict:
    if hasattr(model, "feature_importances_"):
        fi = dict(zip(feature_names, model.feature_importances_))
    elif hasattr(model, "coef_"):
        fi = dict(zip(feature_names, np.abs(model.coef_[0])))
    else:
        return {}
    return {k: round(float(v), 4) for k, v in sorted(fi.items(), key=lambda x: x[1], reverse=True)}


def _risk_label(prob: float) -> str:
    if prob >= 0.75: return "Critical"
    if prob >= 0.50: return "High"
    if prob >= 0.25: return "Medium"
    return "Low"


@router.get("/train")
def train_model(
    filename:   str   = Query(DEFAULT_FILE),
    model_type: str   = Query("random_forest", description="random_forest | logistic_regression | xgboost"),
    test_size:  float = Query(0.2, ge=0.1, le=0.4),
    top_n_employees: int = Query(20, ge=5, le=100, description="Employees to return in per-employee table"),
):
    """
    Train a turnover classifier and return full evaluation + per-employee attrition probabilities.
    """
    df     = load_csv(filename)
    df2, log = _preprocess(df)

    X = df2.drop(columns=[TARGET])
    y = df2[TARGET]

    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=test_size, random_state=42, stratify=y)

    model = _build_model(model_type)
    model.fit(X_tr, y_tr)

    y_pred = model.predict(X_te)
    y_prob = model.predict_proba(X_te)[:, 1]

    acc  = round(accuracy_score(y_te, y_pred), 4)
    prec = round(precision_score(y_te, y_pred, zero_division=0), 4)
    rec  = round(recall_score(y_te, y_pred, zero_division=0), 4)
    f1   = round(f1_score(y_te, y_pred, zero_division=0), 4)
    auc  = round(roc_auc_score(y_te, y_prob), 4)
    cv5  = round(cross_val_score(model, X, y, cv=5, scoring="roc_auc").mean(), 4)

    cm   = confusion_matrix(y_te, y_pred).tolist()
    fi   = _feature_importance(model, list(X.columns))

    # ── Per-employee risk table ────────────────────────────────
    all_probs  = model.predict_proba(X)[:, 1]
    emp_table  = []
    for idx, (row_idx, row) in enumerate(df2.iterrows()):
        prob = float(all_probs[idx])
        raw  = df.iloc[idx]
        emp_table.append({
            "index":            int(row_idx),
            "job_role":         str(raw.get("job_role", "—")),
            "salary":           str(raw.get("salary", "—")),
            "satisfaction_level": float(raw.get("satisfaction_level", 0)),
            "average_montly_hours": int(raw.get("average_montly_hours", 0)),
            "time_spend_clinic": int(raw.get("time_spend_clinic", 0)),
            "actual_left":      int(df2.loc[row_idx, TARGET]),
            "attrition_probability": round(prob, 4),
            "risk_label":       _risk_label(prob),
        })

    emp_table.sort(key=lambda x: x["attrition_probability"], reverse=True)

    # Segment counts
    seg = {
        "critical": sum(1 for e in emp_table if e["risk_label"] == "Critical"),
        "high":     sum(1 for e in emp_table if e["risk_label"] == "High"),
        "medium":   sum(1 for e in emp_table if e["risk_label"] == "Medium"),
        "low":      sum(1 for e in emp_table if e["risk_label"] == "Low"),
    }

    # Attrition probability distribution (histogram)
    hist_counts, hist_edges = np.histogram([e["attrition_probability"] for e in emp_table], bins=10, range=(0, 1))
    prob_dist = {f"{hist_edges[i]:.1f}–{hist_edges[i+1]:.1f}": int(hist_counts[i]) for i in range(len(hist_counts))}

    # Feature-level insights
    top_feat     = next(iter(fi))
    pos_features = [k for k, v in fi.items() if k in ("satisfaction_level",)] if "satisfaction_level" in fi else []
    insights = [
        f"Top turnover predictor: '{top_feat}' (importance {fi[top_feat]}).",
        f"{seg['critical']} employees have Critical attrition probability (≥75%).",
        f"Low satisfaction drives turnover — employees who left had avg satisfaction {df[df[TARGET]==1]['satisfaction_level'].mean():.2f} vs {df[df[TARGET]==0]['satisfaction_level'].mean():.2f} for stayers.",
        f"Overtime correlates with turnover: {df[df['average_montly_hours']>250][TARGET].mean()*100:.1f}% leave rate for >250 h/month.",
        f"Model 5-fold CV AUC: {cv5} — {'Excellent' if cv5>0.95 else 'Good' if cv5>0.85 else 'Moderate'} generalization.",
    ]

    top_k = emp_table[:top_n_employees]
    crit  = [e for e in emp_table if e["risk_label"] == "Critical"][:top_n_employees]

    return {
        "model":          model_type,
        "preprocessing":  log,
        "metrics": {
            "accuracy":    acc,  "precision": prec,
            "recall":      rec,  "f1_score":  f1,
            "roc_auc":     auc,  "cv5_auc":   cv5,
            "train_size":  len(X_tr), "test_size": len(X_te),
        },
        "confusion_matrix": {
            "matrix": cm,
            "true_negatives":  cm[0][0], "false_positives": cm[0][1],
            "false_negatives": cm[1][0], "true_positives":  cm[1][1],
        },
        "feature_importance":     fi,
        "risk_segments":          seg,
        "attrition_prob_distribution": prob_dist,
        "top_risk_employees":     top_k,
        "critical_employees":     crit,
        "auto_insights":          insights,
        "retention_recommendations": [
            "Address low job satisfaction through role enrichment, recognition programs and regular 1-on-1s.",
            "Cap overtime — employees logging >250 h/month show significantly higher turnover.",
            "Create promotion pathways — employees without promotions in 5+ years are high-risk.",
            "Review salary benchmarks — low-salary employees leave at 3× the rate of high-salary peers.",
            "Focus retention efforts on employees in their 2nd–4th year — peak churn window.",
        ],
    }


@router.get("/compare-models")
def compare_models(
    filename:  str   = Query(DEFAULT_FILE),
    test_size: float = Query(0.2, ge=0.1, le=0.4),
):
    """Train all three models and return side-by-side comparison."""
    df     = load_csv(filename)
    df2, _ = _preprocess(df)
    X = df2.drop(columns=[TARGET])
    y = df2[TARGET]
    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=test_size, random_state=42, stratify=y)

    results = {}
    for name in ["logistic_regression", "random_forest", "xgboost"]:
        m = _build_model(name)
        m.fit(X_tr, y_tr)
        yp  = m.predict(X_te)
        ypr = m.predict_proba(X_te)[:, 1]
        cv5 = round(cross_val_score(m, X, y, cv=5, scoring="roc_auc").mean(), 4)
        results[name] = {
            "accuracy":  round(accuracy_score(y_te, yp), 4),
            "roc_auc":   round(roc_auc_score(y_te, ypr), 4),
            "f1_score":  round(f1_score(y_te, yp, zero_division=0), 4),
            "cv5_auc":   cv5,
            "top_feature": next(iter(_feature_importance(m, list(X.columns))), "n/a"),
        }

    best = max(results, key=lambda k: results[k]["roc_auc"])
    return {
        "models":             results,
        "best_model":         best,
        "best_model_auc":     results[best]["roc_auc"],
        "recommendation":     f"Use {best.replace('_',' ').title()} for production — highest ROC AUC ({results[best]['roc_auc']}).",
    }


@router.get("/eda")
def eda(filename: str = Query(DEFAULT_FILE)):
    df   = load_csv(filename)
    rate = round(float(df[TARGET].mean()) * 100, 2)

    role_col  = [c for c in df.columns if "role" in c.lower()][0] if any("role" in c.lower() for c in df.columns) else None
    sat_by_role, turn_by_role, sal_by_role = {}, {}, {}
    if role_col:
        sat_by_role  = {str(k): round(float(v), 3) for k, v in df.groupby(role_col)["satisfaction_level"].mean().items()}
        turn_by_role = {str(k): round(float(v)*100, 2) for k, v in df.groupby(role_col)[TARGET].mean().items()}
        if "salary" in df.columns:
            sal_by_role = {str(k): str(df.groupby(role_col)["salary"].agg(lambda x: x.mode()[0]).get(k, "—")) for k in sat_by_role}

    turn_by_sal = {}
    if "salary" in df.columns:
        turn_by_sal = {str(k): round(float(v)*100, 2) for k, v in df.groupby("salary")[TARGET].mean().items()}

    left_sat  = round(float(df[df[TARGET]==1]["satisfaction_level"].mean()), 3)
    stay_sat  = round(float(df[df[TARGET]==0]["satisfaction_level"].mean()), 3)
    tenure_col = [c for c in df.columns if "time" in c.lower() or "spend" in c.lower()]
    avg_tenure = round(float(df[df[TARGET]==1][tenure_col[0]].mean()), 2) if tenure_col else None

    df2, _ = _preprocess(df)
    corr = df2.corr()[TARGET].drop(TARGET, errors="ignore").sort_values(ascending=False)
    top_pos = {str(k): round(float(v), 4) for k, v in corr[corr > 0].head(5).items()}
    top_neg = {str(k): round(float(v), 4) for k, v in corr[corr < 0].head(5).items()}

    return {
        "dataset_info":               df_info(df),
        "overall_turnover_rate_pct":  rate,
        "avg_satisfaction_who_left":  left_sat,
        "avg_satisfaction_who_stayed": stay_sat,
        "avg_tenure_years_who_left":  avg_tenure,
        "satisfaction_by_role":       sat_by_role,
        "turnover_rate_by_role_pct":  turn_by_role,
        "salary_mode_by_role":        sal_by_role,
        "turnover_rate_by_salary_pct": turn_by_sal,
        "top_positive_correlates_with_turnover": top_pos,
        "top_negative_correlates_with_turnover": top_neg,
        "insight": f"{rate}% overall turnover. Low satisfaction ({left_sat} vs {stay_sat}) is the clearest differentiator.",
    }


@router.get("/risk-profile")
def risk_profile(filename: str = Query(DEFAULT_FILE)):
    df = load_csv(filename)
    results = {}
    for col in df.select_dtypes(include="object").columns:
        g = df.groupby(col)[TARGET].mean().sort_values(ascending=False)
        results[col] = {str(k): round(float(v)*100, 1) for k, v in g.items()}
    if "satisfaction_level" in df.columns:
        df["_sq"] = pd.qcut(df["satisfaction_level"], q=4, labels=["Q1 (low)","Q2","Q3","Q4 (high)"])
        sq = df.groupby("_sq", observed=True)[TARGET].mean()
        results["satisfaction_quartile_pct"] = {str(k): round(float(v)*100, 1) for k, v in sq.items()}
    return {
        "overall_turnover_pct":       round(float(df[TARGET].mean())*100, 2),
        "turnover_risk_by_segment":   results,
    }
